### PR TITLE
Feature/ HICRA implementation

### PR DESCRIPTION
**Pull request by w601sxs**
# What does this PR do?

This branch implements a new trainer that enables language models to develop emergent hierarchical reasoning capabilities through reinforcement
   learning. HICRA is based on research from TIGER-AI-Lab (https://huggingface.co/papers/2509.03646).

  Key Changes

  New Components:
  - HICRATrainer (trl/trainer/hicra_trainer.py:1) - Main trainer extending GRPO
  - HICRAConfig (trl/trainer/hicra_config.py:21) - Configuration class for HICRA-specific parameters
  - Strategic Grams module (trl/trainer/strategic_grams.py:1) - Identifies high-level planning tokens
  - Comprehensive documentation (docs/source/hicra_trainer.md:1)

  Example Scripts:
  - Training example (examples/scripts/hicra_training.py:1)
  - Strategic gram extraction utility (examples/scripts/extract_strategic_grams.py:1)

  Test Coverage:
  - 920 lines of trainer tests (tests/test_hicra_trainer.py:1)
  - 636 lines of validation tests (tests/test_hicra_validation.py:1)

  How It Works

  HICRA amplifies the learning signal for "Strategic Grams" (planning tokens like "let's try a different approach" or "the key insight is"),
  which enables models to:
  1. Separate high-level strategic planning from low-level execution
  2. Learn hierarchical reasoning patterns more efficiently than standard GRPO
  3. Achieve better performance on mathematical reasoning tasks

  The implementation follows the VeRL reference from TIGER-AI-Lab and extends the existing GRPO trainer with hierarchy-aware advantage
  modification.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Adds a new RL trainer that overrides GRPO loss/advantage handling and introduces token-level advantage amplification, which could affect training correctness/performance if edge cases slip through, though changes are largely additive and well-tested.
> 
> **Overview**
> Adds **HICRA (Hierarchy-Aware Credit Assignment)** as a new trainer/config pair (`HICRATrainer`/`HICRAConfig`) that extends GRPO by *amplifying per-token advantages* for high-entropy tokens and optional Strategic-Gram “planning” tokens, including extra HICRA metrics logging and a cached-forward-pass loss path to avoid duplicate model passes.
> 
> Introduces Strategic-Gram utilities (presets, file I/O, extraction helper) plus new docs, examples, and extensive unit/validation tests; updates public exports, docs toctree/paper index, and README to surface the new trainer, and tweaks `.gitignore` for agent artifacts.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a8fcd8d1952de8d4ed390c9edfa11e9c910b2ace. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by w601sxs** [comment]
@qgallouedec or others any comment on this? we just want to add the HICRA algorithm to trl...

File: trl/trainer/hicra_trainer.py [hunk 1]
```diff
@@ -0,0 +1,869 @@
...
+
+            # VeRL: is_higher_entropy = normalized_entropys[is_current_group]>0
+            # VeRL: is_higher_entropy[~should_amplify] = False
+            is_high_entropy = normalized_entropies[is_current_group] > 0  # [group_size, seq_len]
+            is_high_entropy[~should_amplify] = False
+
+            # VeRL: if planning_token_mask is None
+            if planning_token_mask is None:
+                # VeRL: advantages[is_current_group][is_higher_entropy] *= 1 + target_scaler
+                advantages[is_current_group][is_high_entropy] *= 1 + alpha
```
[the remark this answers is not in the document]
Comment (w601sxs) [reply trl/trainer/hicra_trainer.py]: Fixed by building a full 2D mask and using single-level __setitem__ assignment.

File: trl/trainer/hicra_trainer.py [hunk 2]
```diff
@@ -0,0 +1,869 @@
...
+            loss = (per_token_loss * mask).sum() / mask.sum().clamp(min=1.0)
+            loss = loss / self.current_gradient_accumulation_steps
+        elif self.loss_type == "dr_grpo":
+            loss = (per_token_loss * mask).sum() / (per_token_loss.size(0) * self.max_completion_length)
+            loss = loss / self.current_gradient_accumulation_steps
+        elif self.loss_type in ["cispo", "dapo"]:
+            normalizer = inputs["num_items_in_batch"] / self.accelerator.num_processes
+            loss = (per_token_loss * mask).sum() / normalizer
+        else:
+            raise ValueError(f"Unknown loss type: {self.loss_type}")
```
[the remark this answers is not in the document]
Comment (w601sxs) [reply trl/trainer/hicra_trainer.py]: Removed the ~180-line copy-paste of GRPOTrainer._compute_loss. Replaced with a lean override that:                                               
  1. Runs the forward pass once to get entropies                                                                                                   
  2. Modifies advantages (HICRA logic)                                                                                                             
  3. Caches the results on self via a new _get_per_token_logps_and_entropies override                                                              
  4. Calls super()._compute_loss() — which immediately gets the cached values, no second forward pass

File: docs/source/hicra_trainer.md [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+## Citation
+
+```bibtex
+@article{hicra2025,
+  title={Emergent Hierarchical Reasoning in LLMs through Reinforcement Learning},
+  author={TIGER-AI-Lab},
+  journal={arXiv preprint arXiv:2509.03646},
+  year={2025}
+}
+```
```
[the remark this answers is not in the document]
Comment (w601sxs) [reply docs/source/hicra_trainer.md:337]: Added a HICRA subsection under the GRPO section in docs/source/paper_index.md.

**Comment by w601sxs** [comment]
**Fix 1** — Chained boolean indexing silently a no-op                                                                                         
  advantages[is_current_group][is_high_entropy] *= 1 + alpha — PyTorch fancy indexing returns a copy, so the *= was modifying a temp tensor that   
  got discarded. Fixed by building a full 2D mask and using single-level __setitem__ assignment.                                                   
                                                                                                                                                   
**Fix 2** — Duplicated _compute_loss diverging from parent                                                                                    
  Removed the ~180-line copy-paste of GRPOTrainer._compute_loss. Replaced with a lean override that:                                               
  1. Runs the forward pass once to get entropies                                                                                                   
  2. Modifies advantages (HICRA logic)                                                                                                             
  3. Caches the results on self via a new _get_per_token_logps_and_entropies override                                                              
  4. Calls super()._compute_loss() — which immediately gets the cached values, no second forward pass                                              
                                                                                                                                                   
  All loss computation, metric logging, and clip ratio tracking now stays in the parent and stays in sync with future changes.                     
                                                                                                                                                   
**Fix 3**  — Missing paper_index.md entry                                                                                                    
  Added a HICRA subsection under the GRPO section in docs/source/paper_index.md.                                                                   
                                                                                                                                                   
**Note**: Added HICRAConfig/HICRATrainer exports to trl/__init__.py (they were missing, causing all tests to fail with ImportError).

**Comment by w601sxs** [comment]
Both issues fixed and pushed:                                                                                                              
                                                                                   
  1. Added hicra_config/hicra_trainer entries to trl/trainer/__init__.py's _import_structure dict and TYPE_CHECKING block. Without these,   
  the lazy module can't resolve HICRAConfig/HICRATrainer at runtime.                                                                               
  2. Initialized _hicra_cached_logps and _hicra_cached_entropies to None in __init__ and replaced getattr(self, "_hicra_cached_logps", None) 
  with a direct self._hicra_cached_logps is not None check.

File: trl/trainer/hicra_trainer.py [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+from .grpo_trainer import GRPOTrainer
+from .hicra_config import HICRAConfig
+from .strategic_grams import get_default_strategic_grams, load_strategic_grams_from_file
+from .utils import nanmax, nanmin
+
+
+logger = get_logger(__name__)
+
+
+class HICRATrainer(GRPOTrainer):
```
[the remark this answers is not in the document]
Comment (w601sxs) [reply trl/trainer/hicra_trainer.py:54]: HICRA is a adaptation of GRPO so it makes sense to inherit form GRPO trainer.

**Comment by w601sxs** [comment]
@qgallouedec can you or anyone else on the team check this one? 

paper - https://arxiv.org/pdf/2509.03646
original implementation - https://github.com/TIGER-AI-Lab/Hierarchical-Reasoner

File: docs/source/_toctree.yml [hunk 5]
```diff
@@ -17,6 +17,8 @@
     title: DPO
   - local: grpo_trainer
     title: GRPO
+  - local: hicra_trainer
+    title: HICRA
```
[the remark this answers is not in the document]
Comment (w601sxs) [reply docs/source/_toctree.yml:23]: can someone manually check? IMO Cursor bot has a false positive here; the HICRA entry is properly nested 
  as a child of sections just like dpo_trainer, grpo_trainer, etc. No fix needed.

**Comment by w601sxs** [comment]
@albertvillanova @rtrompier @qgallouedec or anyone who can check this?

**Comment by albertvillanova** [comment]
Thanks for the effort here @w601sxs, and apologies for the slow review — this sat in our backlog. 

HICRA is a legitimate paper (TIGER-AI-Lab), but we don't yet see demand for it in TRL (no issues requesting it), and a new ~3.8k-line trainer is a large maintenance surface to adopt speculatively. The design also doesn't match TRL's conventions: a `use_hicra=False` switch inside a dedicated HICRA trainer, caching logps/entropies on `self` to override `_get_per_token_logps_and_entropies`, and hardcoded English "Strategic Gram" phrase matching that won't generalize across languages/tokenizers (the citation block also lists the wrong authors). The branch is now conflicting with `main`, so I'm going to close this for now. If you'd like to pursue it, the best next step is a short feature-request issue to gauge interest. Thanks again!
